vulnerability/events_creator.go

```diff
@@ -210,23 +211,61 @@ func getIdentifierType(id string) string {
 	return s[0]
 }
 
-func getCVSSScore(vul trivy_types.DetectedVulnerability, source db_types.SourceID) float64 {
-	if cvss, ok := vul.CVSS[source]; ok {
-		return cvss.V3Score
+func (e EventsCreator) getCVSSVersion(vul trivy_types.DetectedVulnerability) string {
+	v := getCVSSVector(vul)
+	if v == "" {
+		e.log.Warnf("No CVSS vector found for vulnerability: %s", vul.VulnerabilityID)
+		return ""
+	}
+
+	parts := strings.Split(v, "/")
+	for _, part := range parts {
+		if strings.HasPrefix(part, vectorHeader) {
+			// Extract the number after "CVSS:"
+			version := strings.TrimPrefix(part, vectorHeader)
+			return version
+		}
+	}
+
+	e.log.Warnf("no CVSS version found in vector: %s", v)
+	return ""
+}
+
+func getCVSSValue[T comparable](vul trivy_types.DetectedVulnerability, value func(cvss db_types.CVSS) T, zeroVal T) T {
+	for _, source := range sources {
+		if cvss, ok := vul.CVSS[source]; ok && value(cvss) != zeroVal {
+			return value(cvss)
+		}
+	}
+
+	return zeroVal
+}
+
+func getCVSSVector(vul trivy_types.DetectedVulnerability) string {
+	var zeroVal string
+	if v := getCVSSValue(vul, func(cvss db_types.CVSS) string { return cvss.V3Vector }, zeroVal); v != zeroVal {
+		return v
 	}
+	return getCVSSValue(vul, func(cvss db_types.CVSS) string { return cvss.V2Vector }, zeroVal)
+}
 
-	return 0
+func (e EventsCreator) getCVSSScore(vul trivy_types.DetectedVulnerability) float64 {
```

Comment:
Does this have to be scoped to `(e EventsCreator)`?
If yes, then what about other methods?

Reply:
Yes, as I want it to have the logger.
I've changed other functions (besides the generic one) to methods as well.